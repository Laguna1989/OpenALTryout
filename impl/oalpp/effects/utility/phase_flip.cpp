#include "phase_flip.hpp"

namespace oalpp {
namespace effects {
namespace utility {
float PhaseFlip::process(float input) { return -input; }
void PhaseFlip::reset() { }

} // namespace utility
} // namespace effects
} // namespace oalpp