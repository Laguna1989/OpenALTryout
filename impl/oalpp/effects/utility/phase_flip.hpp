#ifndef OPENALPP_EFFECTS_UTILITY_PHASE_FLIP_HPP
#define OPENALPP_EFFECTS_UTILITY_PHASE_FLIP_HPP

#include <oalpp/effects/mono_effect_interface.hpp>

namespace oalpp {
namespace effects {
namespace utility {

class PhaseFlip : public oalpp::effects::MonoEffectInterface {
public:
    std::vector<float> process(std::vector<float> const& input) override;
};

} // namespace utility
} // namespace effects
} // namespace oalpp

#endif
