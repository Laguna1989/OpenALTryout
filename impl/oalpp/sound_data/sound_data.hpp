#ifndef OPENALPP_SOUND_DATA_HPP
#define OPENALPP_SOUND_DATA_HPP

#include <oalpp/sound_data/sound_data_interface.hpp>

namespace oalpp {

class SoundData : public SoundDataInterface {
public:
    SoundData(std::vector<float> const& data, int sampleRate, int numberOfChannels);
    SoundData() = delete;
    ~SoundData() override = default;

    SoundData(SoundData const& data) = delete;
    SoundData(SoundData&& data) noexcept;

    SoundData& operator=(SoundData const&) = delete;
    SoundData& operator=(SoundData&&) noexcept;

    int getNumberOfChannels() const noexcept override;
    int getSampleRate() const noexcept override;
    std::vector<float> const& getSamples() const noexcept override;

private:
    std::vector<float> m_samples {};
    int m_sampleRate { 0 };
    int m_numberOfChannels { 0 };
};

} // namespace oalpp

#endif // OPENALPP_SOUND_DATA_HPP
