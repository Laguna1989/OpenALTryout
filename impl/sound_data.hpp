#ifndef OPENALPP_SOUND_DATA_HPP
#define OPENALPP_SOUND_DATA_HPP

#include "sound_data_interface.hpp"
#include <string>

namespace oalpp {

class SoundData : public SoundDataInterface {
public:
    explicit SoundData(std::string const& fileName);

    int getNumberOfChannels() const override;
    int getSampleRate() const override;
    std::vector<short> const& getSamples() const override;

private:
    std::vector<short> m_samples {};
    int m_sampleRate { 0 };
    int m_channels { 0 };
};

} // namespace oalpp

#endif // OPENALPP_SOUND_DATA_HPP
