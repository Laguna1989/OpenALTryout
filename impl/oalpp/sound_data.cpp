#include "sound_data.hpp"
#include "libnyquist/Decoders.h"
#include <utility>

namespace oalpp {

SoundData::SoundData(std::string const& fileName)
{
    auto fileData = std::make_unique<nqr::AudioData>();
    nqr::NyquistIO loader;
    loader.Load(fileData.get(), fileName);

    m_channels = fileData->channelCount;
    m_sampleRate = fileData->sampleRate;
    m_samples = std::move(fileData->samples);
}

int SoundData::getNumberOfChannels() const { return m_channels; }

int SoundData::getSampleRate() const { return m_sampleRate; }

std::vector<float> const& SoundData::getSamples() const { return m_samples; }

} // namespace oalpp
