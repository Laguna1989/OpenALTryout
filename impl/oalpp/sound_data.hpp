#ifndef OPENALPP_SOUND_DATA_INCLUDE_HPP
#define OPENALPP_SOUND_DATA_INCLUDE_HPP

#include "oalpp/sound_data/sound_data.hpp"
#include "oalpp/sound_data/sound_data_with_effect.hpp"

#endif // OPENALPP_SOUND_DATA_INCLUDE_HPP
