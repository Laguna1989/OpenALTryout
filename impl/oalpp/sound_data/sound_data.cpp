#include "sound_data.hpp"
#include <utility>

namespace oalpp {

SoundData::SoundData(std::vector<float> const& data, int sampleRate, int numberOfChannels)
    : m_samples { data }
    , m_sampleRate { sampleRate }
    , m_numberOfChannels { numberOfChannels }
{
}

SoundData::SoundData(SoundData&& data) noexcept
    : m_samples { std::move(data.m_samples) }
    , m_sampleRate { std::exchange(data.m_sampleRate, 0) }
    , m_numberOfChannels { std::exchange(data.m_numberOfChannels, 0) }
{
}

SoundData& SoundData::operator=(SoundData&& other) noexcept
{
    m_samples = std::move(other.m_samples);
    m_sampleRate = std::exchange(other.m_sampleRate, 0);
    m_numberOfChannels = std::exchange(other.m_numberOfChannels, 0);
    return *this;
}

int SoundData::getNumberOfChannels() const noexcept { return m_numberOfChannels; }

int SoundData::getSampleRate() const noexcept { return m_sampleRate; }

std::vector<float> const& SoundData::getSamples() const noexcept { return m_samples; }

} // namespace oalpp
