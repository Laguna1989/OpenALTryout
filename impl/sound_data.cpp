#include "sound_data.hpp"
#include "libnyquist/Decoders.h"

SoundData::SoundData(std::string const& fileName)
{
    auto fileData = std::make_shared<nqr::AudioData>();
    nqr::NyquistIO loader;
    loader.Load(fileData.get(), fileName);

    m_channels = fileData->channelCount;
    m_sampleRate = fileData->sampleRate;

    std::size_t size = fileData->samples.size();
    if (fileData->channelCount == 2) {
        if (fileData->samples.size() % 4 != 0) {
            // TODO figure out why there seems to be one sample missing when exporting with reaper.
            size += 2;
        }
    }

    m_samples.resize(size);

    std::transform(fileData->samples.begin(), fileData->samples.end(), m_samples.begin(),
        [](auto in) { return static_cast<short>(in * std::numeric_limits<short>::max()); });
}

int SoundData::getNumberOfChannels() const { return m_channels; }

int SoundData::getSampleRate() const { return m_sampleRate; }

std::vector<short> const& SoundData::getSamples() const { return m_samples; }
