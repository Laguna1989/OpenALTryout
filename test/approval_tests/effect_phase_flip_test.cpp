#include "ApprovalTests/ApprovalTests.hpp"
#include <catch2/catch_test_macros.hpp>
#include <oalpp/effects/utility/phase_flip.hpp>
#include <oalpp/sound_data.hpp>

TEST_CASE("phase_flip")
{
    std::string const fileName { "assets/test1.wav" };
    oalpp::effects::utility::PhaseFlip phaseFlip {};

    oalpp::SoundDataBuilder builder;
    auto const soundWithEffect = builder.fromFile(fileName).withEffect(phaseFlip).create();

    ApprovalTests::Approvals::verifyAll(soundWithEffect.getSamples());
}
