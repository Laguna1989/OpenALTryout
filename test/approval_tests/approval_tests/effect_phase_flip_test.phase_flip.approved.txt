version https://git-lfs.github.com/spec/v1
oid sha256:951e8c8d7fe4b5d659ae576cbf715eec5be6eb187123f73fa79fb68b5c8681a9
size 231811
